pageserver/src/bin/pageserver.rs

```diff
@@ -337,33 +337,120 @@ fn start_pageserver(
+    };
+
     // Scan the local 'tenants/' directory and start loading the tenants
     let init_started_at = std::time::Instant::now();
+    let shutdown_pageserver = tokio_util::sync::CancellationToken::new();
+
     BACKGROUND_RUNTIME.block_on(mgr::init_tenant_mgr(
         conf,
         broker_client.clone(),
         remote_storage.clone(),
-        (init_done_tx, init_done_rx.clone()),
+        order,
     ))?;
 
     BACKGROUND_RUNTIME.spawn({
-        let init_done_rx = init_done_rx.clone();
-        async move {
+        let init_done_rx = init_done_rx;
+        let shutdown_pageserver = shutdown_pageserver.clone();
+        let wait_and_report = async move {
+            // NOTE: unlike many futures in pageserver, this one is cancellation-safe
+            let guard = scopeguard::guard_on_success((), |_| tracing::info!("Cancelled before initial load completed"));
+
             init_done_rx.wait().await;
 
-            let elapsed = init_started_at.elapsed();
+            scopeguard::ScopeGuard::into_inner(guard);
+
+            let init_done = std::time::Instant::now();
+            let elapsed = init_done - init_started_at;
 
             tracing::info!(
                 elapsed_millis = elapsed.as_millis(),
-                "Initial load completed."
+                "Initial load completed"
             );
+
+            // no need to do anything here for initial_logical_size_can_start, because it was
+            // waiting on the same barrier as above.
+
+            let mut init_sizes_done = std::pin::pin!(init_logical_size_done_rx.wait());
+
+            // it is difficult to define this timeout. our largest initialization completions are
+            // in the range of 100-200s, so perhaps this works. smaller nodes will have background
+            // tasks "not running" for this long unless every timeline has it's initial logical
+            // size calculated. not running background tasks for some seconds is not terrible.
+            let timeout = conf.background_task_maximum_delay;
+
+            let guard = scopeguard::guard_on_success((), |_| tracing::info!("Cancelled before initial logical sizes completed"));
+
+            let completed = tokio::select! {
+                _ = &mut init_sizes_done => {
+                    let now = std::time::Instant::now();
+                    tracing::info!(
+                        from_init_done_millis = (now - init_done).as_millis(),
+                        from_init_millis = (now - init_started_at).as_millis(),
+                        "Initial logical sizes completed"
+                    );
+                    true
+                }
+                _ = tokio::time::sleep(timeout) => {
+                    tracing::info!(
+                        timeout_millis = timeout.as_millis(),
+                        "Initial logical size timeout elapsed; starting background jobs"
+                    );
+                    false
+                }
+            };
```

Comment:
I've been using this elsewhere as well:

```
let completed = tokio::select! { ... };

if !completed {
    complete
}
```

I mean I guess this could be changed but why?

Reply:
I don't like the boolean true/false => false/true you proposed. Perhaps a better alternative is to return a `Option<&mut {future]>` of the incomplete so we don't have to name this and avoid the true/false false/true altogether.